expression/builtin_convert_charset.go

```diff
@@ -108,28 +127,154 @@ func (b *builtinInternalToBinarySig) vecEvalString(input *chunk.Chunk, result *c
 	return nil
 }
 
-// toBinaryMap contains the builtin functions which arguments need to be converted to the correct charset.
-var toBinaryMap = map[string]struct{}{
-	ast.Hex: {}, ast.Length: {}, ast.OctetLength: {}, ast.ASCII: {},
-	ast.ToBase64: {},
+type tidbFromBinaryFunctionClass struct {
+	baseFunctionClass
+
+	tp *types.FieldType
 }
 
-// WrapWithToBinary wraps `expr` with to_binary sig.
-func WrapWithToBinary(ctx sessionctx.Context, expr Expression, funcName string) Expression {
-	exprTp := expr.GetType()
-	if _, err := charset.GetDefaultCollationLegacy(exprTp.Charset); err != nil {
-		if _, ok := toBinaryMap[funcName]; ok {
-			fc := funcs[InternalFuncToBinary]
-			sig, err := fc.getFunction(ctx, []Expression{expr})
-			if err != nil {
-				return expr
+func (c *tidbFromBinaryFunctionClass) getFunction(ctx sessionctx.Context, args []Expression) (builtinFunc, error) {
+	if err := c.verifyArgs(args); err != nil {
+		return nil, c.verifyArgs(args)
+	}
+	argTp := args[0].GetType().EvalType()
+	var sig builtinFunc
+	switch argTp {
+	case types.ETString:
+		bf, err := newBaseBuiltinFuncWithTp(ctx, c.funcName, args, types.ETString, types.ETString)
+		if err != nil {
+			return nil, err
+		}
+		bf.tp = c.tp
+		sig = &builtinInternalFromBinarySig{bf}
+		sig.setPbCode(tipb.ScalarFuncSig_FromBinary)
+	default:
+		return nil, fmt.Errorf("unexpected argTp: %d", argTp)
+	}
+	return sig, nil
+}
+
+type builtinInternalFromBinarySig struct {
+	baseBuiltinFunc
+}
+
+func (b *builtinInternalFromBinarySig) Clone() builtinFunc {
+	newSig := &builtinInternalFromBinarySig{}
+	newSig.cloneFrom(&b.baseBuiltinFunc)
+	return newSig
+}
+
+func (b *builtinInternalFromBinarySig) evalString(row chunk.Row) (res string, isNull bool, err error) {
+	val, isNull, err := b.args[0].EvalString(b.ctx, row)
+	if isNull || err != nil {
+		return val, isNull, err
+	}
+	transferString := b.getTransferFunc()
+	tBytes, err := transferString([]byte(val))
+	return string(tBytes), false, err
+}
+
+func (b *builtinInternalFromBinarySig) vectorized() bool {
+	return true
+}
+
+func (b *builtinInternalFromBinarySig) vecEvalString(input *chunk.Chunk, result *chunk.Column) error {
+	n := input.NumRows()
+	buf, err := b.bufAllocator.get()
+	if err != nil {
+		return err
+	}
+	defer b.bufAllocator.put(buf)
+	if err := b.args[0].VecEvalString(b.ctx, input, buf); err != nil {
+		return err
+	}
+	transferString := b.getTransferFunc()
+	result.ReserveString(n)
+	for i := 0; i < n; i++ {
+		if buf.IsNull(i) {
+			result.AppendNull()
+			continue
+		}
+		str, err := transferString(buf.GetBytes(i))
```

Comment:
We can use a buffer to reduce memory allocation.

Reply:
yes, it has the buffer, see https://github.com/pingcap/tidb/pull/29905/files#diff-b7c01f9c9ab46f762c72934562d5e01b7ec3750c1b6b6601147b5b2808d30d5cR218